### PR TITLE
Admin Router: Redundant slashes test for `/service` endpoint

### PR DESCRIPTION
**Pull request by vespian**
## High Level Description

This PR introduces a test that checks if duplicate slashes are handled by AR.

The question here is whether we want to support `/service` endpoint mapping request for `/service/foo///bar/baz.jpg` to a different task/framework than a request for `/service/foo/bar/baz.jpg`. 

The RFC says https://www.rfc-editor.org/rfc/rfc3986.txt (chapter 3.3):

```
A path consists of a sequence of path segments separated by a slash ("/") character.
```
instead of
```
The path may consist of a sequence of path segments separated by a single slash "/" character.
```

So at least in theory, it is valid to have empty URL segments in the path and they are treated as valid segments which take part in routing decision.  On the other hand, Apache HTTPd and Linux file systems map multiple adjacent segments to the same path (https://webmasters.stackexchange.com/a/8381).

Let's look at this from the perspective of DC/OS:
* what is the current behaviour? 
Application ID consists of Application Groups. I have made a simple experiment - installed a service into Application id `foo/bar///baz/Nginx` and then tried with `foo/bar/baz/Nginx`. The second attempt failed with the message that service with given Application ID already exists. So it seems that Marathon at least internally does not differentiate between these two. AR in this situation was still able to route to correct destination as the request path and Application ID are both normalized (-> redunant slashes are removed) before the routing decision is made. In case of Mesos I think this works a bit more transparently as frameworkID field in `/steate-summary` output is ATM free-form field and and you can have both duplicate slashes in there as well as multiple frameworks registering with the same Application ID.
* what is the "bigger picuture" when it comes to DC/OS and its principles
I am not 100% sure about exact DC/OS terminology but Application Groups are basically namespaces. Having empty Application Group (-> duplicate slashes) means that we have a namespace that is named "" (empty string). At least in bouncer we were forcing groups to be at least few characters long, I am not sure if such a namespace names make sense.

So my conclusion was to follow the "robustness principle": https://en.wikipedia.org/wiki/Robustness_principle#Interpretation
```
RFC 1122 (1989) expanded on Postel's principle by recommending that programmers "assume that the network is filled with malevolent entities that will send in packets designed to have the worst possible effect".[2] Protocols should allow for the addition of new codes for existing fields in future versions of protocols by accepting messages with unknown codes (possibly logging them). Programmers should avoid sending messages with "legal but obscure protocol features" that might expose deficiencies in receivers, and design their code "not just to survive other misbehaving hosts, but also to cooperate to limit the amount of disruption such hosts can cause to the shared communication facility".[3]
```
AR currently normalizes all paths into a common representation before comparing them. This is compatible with the assumption that empty namespace name is an invalid name.

As for the reasoning behind this test in general - I had a discussion with Jose ~2 months ago, he had a problem with a client sending duplicate slashes and thus breaking Cosmos. Writing this test helped me convince him that this is unrelated to Admin Router. I can find this conversation if necessary but it will take time.

## Related issues
https://jira.mesosphere.com/browse/DCOS_OSS-1510 Admin Router: Write tests for AR behaviour wrt. handling duplicate slashes by `/service` endpoint

## Related PRs:
AR-NEXT: https://github.com/dcos/dcos/pull/1802

### CODE REVIEW

File: packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml
Comment (adamtheturtle) [inline packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml]: Is this still necessary?

File: packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml
Comment (adamtheturtle) [inline packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml]: It isn't clear to me what this is testing. Please add a comment explaining what the purpose of this test is. If you feel that nothing could be added that isn't said in the test, ping me on Slack.

**Review by adamtheturtle — CHANGES_REQUESTED** [review]
Thanks @vespian

File: packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml
Comment (vespian) [reply packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml]: Good catch!

No, it is - it is just an artefact of rebasing/moving commits around. Fixed!

**Comment by vespian** [comment]
@adamtheturtle Done, ready for another pass.

File: packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml
Comment (adamtheturtle) [inline packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml]: Somehow change `This test checks tests if` - I guess there was a typo.

File: packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml
Comment (adamtheturtle) [inline packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml]: `s/correct/the correct`

File: packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml
Comment (adamtheturtle) [inline packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml]: `s/lets/let's`.

**Review by adamtheturtle — APPROVED** [review]
Adding a test pretty much adds a burden of support for future maintainers. Given the spec, this LGTM modulo minor changes. However, I would like this reviewed by someone ( @jgehrcke ?) with the authority to approve this.

**Comment by vespian** [comment]
@adamtheturtle All fixed, @jgehrcke please have a look so that we can 🚢

File: packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml
Comment (adamtheturtle) [inline packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml:79]: ❓ If we could find this RFC / confirm, that would be ideal.

Comment (adamtheturtle) [reply packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml:79]: Also "what is 'the correct service'"? is a question that comes to mind. Is it possible that someone would want a different endpoint at `baz//bat` than `baz/bat`?

Comment (adamtheturtle) [reply packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml:79]: I guess you're talking about:

>  The path may consist of a sequence of path segments separated by a
   single slash "/" character.

https://www.ietf.org/rfc/rfc2396.txt

Comment (adamtheturtle) [reply packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml:79]: Attempting to @jgehrcke with the above: Let's be precise here.

In particular if this is behaviour not supported by other systems or standards, this is exactly the kind of test that might break in a big rewrite. We want the maintainer at that point to be able to make an informed decision on whether or not to keep the test.

That decision may be based on:
* To what extent do people rely on this functionality - we may never know
* Is this required behaviour to meet standards we aim or claim to meet
* Is this required behaviour to match behaviour of other tools that we aim to match

This comment can try to address some of this.

> To what extent do people rely on this functionality

Why are we adding this test? We add tests in order to provide regression-safety for functionality we do not want to accidentally disappear. Why is this functionality in this category? Is it because someone requested it?

> Is this required behaviour to meet standards we aim or claim to meet

I'm not sure but perhaps that RFC will give us a clue.
Actually the double slash, from what I can gather should *not* be considered the same as a single slash. In fact a double slash should indicate an empty path.
See https://stackoverflow.com/a/40683096.

> Is this required behaviour to match behaviour of other tools that we aim to match

Maybe?

I see

>  in (most?) modern operating systems (Linux/Unix, Windows), multiple path separators in a row do not have any special meaning, so /path/to/foo and /path//to////foo would eventually map to the same file. - https://stackoverflow.com/a/10161264

and this adds to the idea that this is behaviour we should support.

**Review by adamtheturtle — DISMISSED** [review]
Sorry I'm retracting my approval after more thinking as per my most recent comments.

File: packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml
Comment (vespian) [reply packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml:79]: The question here is whether we want to support `/service` endpoint mapping request for `/service/foo///bar/baz.jpg` to a different task/framework than a request for `/service/foo/bar/baz.jpg`. 

The RFC that Adam linked (https://www.ietf.org/rfc/rfc2396.txt) has been obsoleted by https://www.rfc-editor.org/rfc/rfc3986.txt (chapter 3.3):

```
A path consists of a sequence of path segments separated by a slash ("/") character.
```
instead of
```
The path may consist of a sequence of path segments separated by a single slash "/" character.
```

So at least in theory, it is valid to have empty URL segments in the path and they are treated as valid segments which take part in routing decision.  On the other hand, Apache HTTPd and Linux file systems map multiple adjacent segments to the same path (https://webmasters.stackexchange.com/a/8381).

Let's look at this from the perspective of DC/OS:
* what is the current behaviour? 
Application ID consists of Application Groups. I have made a simple experiment - installed a service into Application id `foo/bar///baz/Nginx` and then tried with `foo/bar/baz/Nginx`. The second attempt failed with the message that service with given Application ID already exists. So it seems that Marathon at least internally does not differentiate between these two. AR in this situation was still able to route to correct destination as the request path and Application ID are both normalized (-> redunant slashes are removed) before the routing decision is made. In case of Mesos I think this works a bit more transparently as frameworkID field in `/steate-summary` output is ATM free-form field and and you can have both duplicate slashes in there as well as multiple frameworks registering with the same Application ID.
* I am not 100% sure about DC/OS terminology but Application Groups are basically namespaces. Having empty Application Group (-> duplicate slashes) means that we have a namespace that is named "" (empty string). At least in bouncer we were forcing groups to be at least few characters long, I am not sure if such namespace names make sense.

So my conclusion was to follow the "robustness principle": https://en.wikipedia.org/wiki/Robustness_principle#Interpretation
```
RFC 1122 (1989) expanded on Postel's principle by recommending that programmers "assume that the network is filled with malevolent entities that will send in packets designed to have the worst possible effect".[2] Protocols should allow for the addition of new codes for existing fields in future versions of protocols by accepting messages with unknown codes (possibly logging them). Programmers should avoid sending messages with "legal but obscure protocol features" that might expose deficiencies in receivers, and design their code "not just to survive other misbehaving hosts, but also to cooperate to limit the amount of disruption such hosts can cause to the shared communication facility".[3]
```
So AR currently normalizes all paths into a common representation before comparing them. This is compatible with the assumption that empty namespace name is an invalid name.

As for the reasoning behind this test in general - I had a discussion with Jose ~2 months ago, he had a problem with a client sending duplicate slashes and thus breaking Cosmos. Writing this test helped me convince him that this is unrelated to Admin Router. I can find this conversation if necessary.

Comment (vespian) [reply packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml:79]: I just went a little bit deeper, it turns out that deploying Marathon Service into Application ID `foo/bar///baz/nginx` resulted in `/marathon/v2/apps` output:
```
{

    "apps": [
        {
            "id": "/foo/bar/baz/nginx",
            "acceptedResourceRoles": [
                "*"
            ],
            "backoffFactor": 1.15,
            "backoffSeconds": 1,
            "container": {
                "type": "DOCKER",
                "docker": {
                    "forcePullImage": false,
                    "image": "bitnami/nginx:1.10.3-r1",
                    "parameters": [ ],
                    "privileged": false
                },
                "volumes": [ ],
                "portMappings": [
                    {
                        "containerPort": 80,
                        "hostPort": 0,
                        "labels": { },
                        "protocol": "tcp",
                        "servicePort": 10000
                    },
                    {
                        "containerPort": 443,
                        "hostPort": 0,
                        "labels": { },
                        "protocol": "tcp",
                        "servicePort": 10001
                    }
                ]
            },
```

So internally, Marathon also normalizes Application ID/removes slashes and current AR behaviour is consistent with Marathon's.

Thanks @adamtheturtle for going so deep and for being an awsome reviewer, this was an interesting finding.

**Comment by jgehrcke** [comment]
Is there a JIRA ticket that tracks this?

**Comment by vespian** [comment]
> Is there a JIRA ticket that tracks this?

You mean adding this test - no, there is not. This was a quick ad-hoc action to prove Jose that his problem is unrelated to AR. Adding this test took 10 minutes.

Do you think we should create a Jira for that?

**Review by gpaul — APPROVED** [review]
Gets my vote for PR of the month. Thanks @vespian and @adamtheturtle for teaching me multiple interesting things during this review.

**Comment by jgehrcke** [comment]
> AR currently normalizes all paths into a common representation before comparing them

Related: http://nginx.org/en/docs/http/ngx_http_core_module.html#merge_slashes (default: *on*).

**Comment by jgehrcke** [comment]
> he had a problem with a client sending duplicate slashes and thus breaking Cosmos

Out of curiosity: did the client send a request URL with multiple slashes as path separator through Admin Router? According to the nginx `merge_slashes` default behavior this should be no different from having a single path separator.

**Review by jgehrcke — APPROVED** [review]
IMO all the required work is done, review state is great.

Let's go ahead with this.

**Comment by vespian** [comment]
> Out of curiosity: did the client send a request URL with multiple slashes as path separator through Admin Router? According to the nginx merge_slashes default behavior this should be no different from having single path separator.

I do not remember, this was a quick fix I did while I was on holidays - noticed him having a problem in one of the channels, jumped in to help, quickly wrote the test to confirm that AR is not to blame. From what I know the issue was later resolved, but I stopped following at some point and have no idea what exactly the problem was.

I can try to find the relevant conversation, but this will take me 30-60 minutes.

**Pull request merged**
